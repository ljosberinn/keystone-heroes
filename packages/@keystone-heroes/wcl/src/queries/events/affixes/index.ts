import { Affixes } from "@keystone-heroes/db/types";

import type { EventParams } from "..";
import type {
  AbsorbEvent,
  AllTrackedEventTypes,
  ApplyBuffEvent,
  ApplyDebuffStackEvent,
  DamageEvent,
  HealEvent,
  InterruptEvent,
} from "../types";
import {
  filterExpression as bolsteringFilterExpression,
  getBolsteringEvents,
} from "./bolstering";
import {
  filterExpression as burstingFilterExpression,
  getBurstingEvents,
} from "./bursting";
import {
  filterExpression as explosiveFilterExpression,
  getExplosiveEvents,
} from "./explosive";
import {
  filterExpression as grievousFilterExpression,
  getGrievousEvents,
} from "./grievous";
import {
  filterExpression as necroticFilterExpression,
  getNecroticEvents,
} from "./necrotic";
import {
  filterExpression as quakingFilterExpression,
  getQuakingEvents,
} from "./quaking";
import {
  filterExpression as sanguineFilterExpression,
  getSanguineEvents,
} from "./sanguine";
import {
  filterExpression as spitefulFilterExpression,
  getSpitefulEvents,
} from "./spiteful";
import {
  filterExpression as stormingFilterExpression,
  getStormingEvents,
} from "./storming";
import {
  filterExpression as tormentedFilterExpression,
  getTormentedEvents,
} from "./tormented";
import {
  filterExpression as volcanicFilterExpression,
  getVolcanicEvents,
} from "./volcanic";

type AffixWithEvents = Exclude<
  Affixes,
  | "Awakened"
  | "Beguiling"
  | "Tyrannical"
  | "Fortified"
  | "Prideful"
  | "Raging"
  | "Infested"
  | "Inspiring"
  | "Reaping"
  | "Skittish"
>;

const affixExpressionMap: Record<AffixWithEvents, string[]> = {
  [Affixes.Sanguine]: sanguineFilterExpression,
  [Affixes.Explosive]: explosiveFilterExpression,
  [Affixes.Grievous]: grievousFilterExpression,
  [Affixes.Necrotic]: necroticFilterExpression,
  [Affixes.Volcanic]: volcanicFilterExpression,
  [Affixes.Bursting]: burstingFilterExpression,
  [Affixes.Spiteful]: spitefulFilterExpression,
  [Affixes.Quaking]: quakingFilterExpression,
  [Affixes.Storming]: stormingFilterExpression,
  [Affixes.Bolstering]: bolsteringFilterExpression,
  [Affixes.Tormented]: tormentedFilterExpression,
};

export const getAffixExpression = (affixes: Affixes[]): string[] => {
  return affixes
    .filter((affix): affix is AffixWithEvents => affix in affixExpressionMap)
    .flatMap((affix) => affixExpressionMap[affix]);
};

export const filterAffixEvents = (
  allEvents: AllTrackedEventTypes[],
  affixes: EventParams["affixes"]
): (
  | AbsorbEvent
  | InterruptEvent
  | HealEvent
  | DamageEvent
  | ApplyBuffEvent
  | ApplyDebuffStackEvent
)[] => {
  const affixSet = new Set(affixes);

  // seasonal
  const tormentedEvents = getTormentedEvents(allEvents, affixSet);

  // common affixes
  const storming = getStormingEvents(allEvents, affixSet);
  const spiteful = getSpitefulEvents(allEvents, affixSet);
  const sanguine = getSanguineEvents(allEvents, affixSet);
  const volcanic = getVolcanicEvents(allEvents, affixSet);
  const quaking = getQuakingEvents(allEvents, affixSet);
  const bolstering = getBolsteringEvents(allEvents, affixSet);
  const bursting = getBurstingEvents(allEvents, affixSet);
  const explosive = getExplosiveEvents(allEvents, affixSet);
  const grievous = getGrievousEvents(allEvents, affixSet);
  const necrotic = getNecroticEvents(allEvents, affixSet);

  return [
    ...tormentedEvents,
    ...storming,
    ...spiteful,
    ...sanguine,
    ...volcanic,
    ...quaking,
    ...bolstering,
    ...bursting,
    ...explosive,
    ...grievous,
    ...necrotic,
  ];
};
