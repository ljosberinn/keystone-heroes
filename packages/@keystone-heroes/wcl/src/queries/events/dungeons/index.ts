import { DungeonIDs } from "@keystone-heroes/db/data";
import type { PlayableClass } from "@keystone-heroes/db/types";

import type {
  AllTrackedEventTypes,
  ApplyBuffEvent,
  ApplyBuffStackEvent,
  ApplyDebuffEvent,
  BeginCastEvent,
  CastEvent,
  DamageEvent,
  DeathEvent,
  HealEvent,
  RemoveBuffEvent,
} from "../../events/types";
import { filterExpression as dosFilterExpression, getDOSEvents } from "./dos";
import { filterExpression as hoaFilterExpression, getHOAEvents } from "./hoa";
import { filterExpression as nwFilterExpression, getNWEvents } from "./nw";
import { filterExpression as pfFilterExpression, getPFEvents } from "./pf";
import { filterExpression as sdFilterExpression, getSDEvents } from "./sd";
import { filterExpression as soaFilterExpression, getSOAEvents } from "./soa";
import { filterExpression as topFilterExpression, getTOPEvents } from "./top";

type DungeonWithEvents = Exclude<DungeonIDs, DungeonIDs.MISTS_OF_TIRNA_SCITHE>;

const isDungeonWithEvent = (id: DungeonIDs): id is DungeonWithEvents =>
  id in dungeonExpressionMap;

const dungeonExpressionMap: Record<DungeonWithEvents, string[]> = {
  [DungeonIDs.DE_OTHER_SIDE]: dosFilterExpression,
  [DungeonIDs.HALLS_OF_ATONEMENT]: hoaFilterExpression,
  [DungeonIDs.PLAGUEFALL]: pfFilterExpression,
  [DungeonIDs.SANGUINE_DEPTHS]: sdFilterExpression,
  [DungeonIDs.SPIRES_OF_ASCENSION]: soaFilterExpression,
  [DungeonIDs.THEATRE_OF_PAIN]: topFilterExpression,
  [DungeonIDs.THE_NECROTIC_WAKE]: nwFilterExpression,
};

export const getDungeonExpression = (id: DungeonIDs): string[] =>
  isDungeonWithEvent(id) ? dungeonExpressionMap[id] : [];

export const filterDungeonEvents = (
  allEvents: AllTrackedEventTypes,
  dungeonID: DungeonIDs,
  playerMetaInformation: { actorID: number; class: PlayableClass }[]
):
  | ApplyDebuffEvent[]
  | CastEvent[]
  | (DeathEvent | DamageEvent | ApplyBuffEvent)[]
  | (BeginCastEvent | ApplyBuffEvent | RemoveBuffEvent | ApplyBuffStackEvent)[]
  | (DamageEvent | HealEvent)[] => {
  switch (dungeonID) {
    case DungeonIDs.DE_OTHER_SIDE:
      return getDOSEvents(allEvents);
    case DungeonIDs.HALLS_OF_ATONEMENT:
      return getHOAEvents(allEvents);
    case DungeonIDs.PLAGUEFALL:
      return getPFEvents(allEvents, playerMetaInformation);
    case DungeonIDs.SANGUINE_DEPTHS:
      return getSDEvents(allEvents);
    case DungeonIDs.SPIRES_OF_ASCENSION:
      return getSOAEvents(allEvents);
    case DungeonIDs.THEATRE_OF_PAIN:
      return getTOPEvents(allEvents);
    case DungeonIDs.THE_NECROTIC_WAKE:
      return getNWEvents(allEvents);
    default:
      return [];
  }
};
