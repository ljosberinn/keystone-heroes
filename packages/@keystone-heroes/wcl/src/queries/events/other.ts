import { remarkableSpellIDs } from "@keystone-heroes/db/data";

import { EventDataType, HostilityType } from "../../types";
import type { GetEventBaseParams } from "./utils";
import { getEvents } from "./utils";

import type {
  CastEvent,
  BeginCastEvent,
  ApplyBuffEvent,
  RemoveBuffEvent,
} from ".";

// TODO: drums

export const getRemarkableSpellCastEvents = async (
  params: GetEventBaseParams,
  playerActorIDs: Set<number>
): Promise<CastEvent[]> => {
  const allEvents = await getEvents<CastEvent | BeginCastEvent>({
    ...params,
    dataType: EventDataType.Casts,
    hostilityType: HostilityType.Friendlies,
  });

  return allEvents.filter((event): event is CastEvent => {
    return (
      event.type === "cast" &&
      playerActorIDs.has(event.sourceID) &&
      remarkableSpellIDs.has(event.abilityGameID)
    );
  });
};

// type CardboardAssassinUsages = { events: CastEvent[]; actorId: number | null };
// export const CARDBOARD_ASSASSIN = 51_229;

// const getCardboardAssassinUsage = async (
//   params: GetEventBaseParams
// ): Promise<CardboardAssassinUsages[]> => {
//   const sdk = await getCachedSdk();

//   const data = await sdk.PetActors({
//     reportID: params.reportID,
//   });

//   const cardboardAssassinInstances =
//     data?.reportData?.report?.masterData?.actors?.filter(
//       (pet) => pet?.gameID === CARDBOARD_ASSASSIN
//     ) ?? [];

//   if (cardboardAssassinInstances.length === 0) {
//     return [];
//   }

//   const eventGroup = await Promise.all(
//     cardboardAssassinInstances.map(async (instance) => {
//       if (!instance?.id) {
//         return null;
//       }

//       const events = await getEvents<CastEvent>({
//         ...params,
//         sourceID: instance.id,
//         dataType: EventDataType.Threat,
//         hostilityType: HostilityType.Friendlies,
//       });

//       if (events.length === 0) {
//         return null;
//       }

//       return {
//         actorId: instance.petOwner,
//         events,
//       };
//     })
//   );

//   return eventGroup.filter(
//     (dataset): dataset is CardboardAssassinUsages => dataset !== null
//   );
// };

export const INVISIBILITY = {
  DIMENSIONAL_SHIFTER: 321_422,
  POTION_OF_THE_HIDDEN_SPIRIT: 307_195,
} as const;

export const getInvisibilityUsage = async (
  params: GetEventBaseParams
): Promise<ApplyBuffEvent[]> => {
  const dimensionalShifterUsage = await getEvents<
    ApplyBuffEvent | RemoveBuffEvent
  >({
    ...params,
    dataType: EventDataType.Buffs,
    hostilityType: HostilityType.Friendlies,
    abilityID: INVISIBILITY.DIMENSIONAL_SHIFTER,
  });

  const potionUsage = await getEvents<ApplyBuffEvent | RemoveBuffEvent>({
    ...params,
    dataType: EventDataType.Buffs,
    hostilityType: HostilityType.Friendlies,
    abilityID: INVISIBILITY.POTION_OF_THE_HIDDEN_SPIRIT,
  });

  return [...dimensionalShifterUsage, ...potionUsage].filter(
    (event): event is ApplyBuffEvent => event.type === "applybuff"
  );
};

export const ENGINEERING_BATTLE_REZ = {
  // Disposable Spectrophasic Reanimator
  SHADOWLANDS: 345_130,
};

export const getEngineeringBattleRezCastEvents = async (
  params: GetEventBaseParams
): Promise<CastEvent[]> => {
  const allEvents = await getEvents<BeginCastEvent | CastEvent>({
    ...params,
    dataType: EventDataType.Casts,
    hostilityType: HostilityType.Friendlies,
    abilityID: ENGINEERING_BATTLE_REZ.SHADOWLANDS,
  });

  return allEvents.filter((event): event is CastEvent => event.type === "cast");
};
