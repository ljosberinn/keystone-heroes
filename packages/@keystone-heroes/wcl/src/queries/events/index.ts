import { DungeonIDs, remarkableSpellIDs } from "@keystone-heroes/db/data";
import { PlayableClass, Affixes } from "@keystone-heroes/db/types";

import { filterProfessionEvents } from "../events/other";
import type {
  AnyEvent,
  ApplyDebuffEvent,
  CastEvent,
  DamageEvent,
  DeathEvent,
} from "../events/types";
import {
  filterExpression as bolsteringFilterExpression,
  getHighestBolsteringStack,
  isBolsteringEvent,
} from "./affixes/bolstering";
import {
  filterExpression as burstingFilterExpression,
  isBurstingEvent,
} from "./affixes/bursting";
import {
  createIsExplosiveDeathEvent,
  filterExpression as explosiveFilterExpression,
  findExplosiveTargetID,
  isExplosiveDamageEvent,
} from "./affixes/explosive";
import {
  filterExpression as grievousFilterExpression,
  isGrievousDamageEvent,
} from "./affixes/grievous";
import {
  filterExpression as necroticFilterExpression,
  getHighestNecroticStack,
  isNecroticDamageEvent,
  isNecroticStackEvent,
} from "./affixes/necrotic";
import {
  filterExpression as quakingFilterExpression,
  isQuakingDamageEvent,
  isQuakingInterruptEvent,
} from "./affixes/quaking";
import {
  filterExpression as sanguineFilterExpression,
  isSanguineDamageEvent,
  isSanguineHealEvent,
  reduceHealingDoneBySanguine,
} from "./affixes/sanguine";
import {
  filterExpression as spitefulFilterExpression,
  isSpitefulDamageEvent,
} from "./affixes/spiteful";
import {
  filterExpression as stormingFilterExpression,
  isStormingEvent,
} from "./affixes/storming";
import {
  filterExpression as tormentedFilterExpression,
  isBitingColdDamageEvent,
  isBottleOfSanguineIchorDamageEvent,
  isBottleOfSanguineIchorHealEvent,
  isColdSnapDamageEvent,
  isCrushDamageEvent,
  isFrostLanceDamageEvent,
  isInfernoDamageEvent,
  isRazeDamageEvent,
  isScorchingBlastDamageEvent,
  isSeismicWaveDamageEvent,
  isSeverDamageEvent,
  isSoulforgeFlamesDamageEvent,
  isStoneWardEvent,
  isStygianKingsBarbsEvent,
  isTheFifthSkullDamageEvent,
  isVolcanicPlumeDamageEvent,
} from "./affixes/tormented";
import {
  filterExpression as volcanicFilterExpression,
  isVolcanicEvent,
} from "./affixes/volcanic";
import {
  filterExpression as dosFilterExpression,
  isDosUrnEvent,
} from "./dungeons/shadowlands/dos";
import {
  filterExpression as hoaFilterExpression,
  isHoaGargoyleEvent,
} from "./dungeons/shadowlands/hoa";
import {
  filterExpression as nwFilterExpression,
  isNwHammerEvent,
  isNwKyrianOrbDamageEvent,
  isNwKyrianOrbHealEvent,
  isNwOrbEvent,
  isNwSpearEvent,
  nwOrbReducer,
  nwSpearReducer,
} from "./dungeons/shadowlands/nw";
import {
  filterExpression as pfFilterExpression,
  isPfPlagueBombDamageEvent,
  isPfSlimeBuffEvent,
  isPfSlimeDeathEvent,
} from "./dungeons/shadowlands/pf";
import {
  filterExpression as sdFilterExpression,
  isSdLanternBuffEvent,
  isSdLanternOpeningEvent,
} from "./dungeons/shadowlands/sd";
import {
  filterExpression as soaFilterExpression,
  isSoaSpearEvent,
  soaSpearReducer,
} from "./dungeons/shadowlands/soa";
import {
  filterExpression as topFilterExpression,
  isTopBannerAuraEvent,
} from "./dungeons/shadowlands/top";
import { recursiveGetEvents, reduceEventsByPlayer } from "./utils";

export * from "./other";
export * from "./types";

export type EventParams = {
  reportID: string;
  startTime: number;
  endTime: number;
  dungeonID: DungeonIDs;
  fightID: number;
  affixes: Affixes[];
};

type DungeonWithEvents = Exclude<DungeonIDs, DungeonIDs.MISTS_OF_TIRNA_SCITHE>;

const isDungeonWithEvent = (id: DungeonIDs): id is DungeonWithEvents =>
  id in dungeonExpressionMap;

// TODO: feign false doesnt work?
const deathFilterExpression =
  'target.type = "player" and type = "death" and feign = false';
const remarkableSpellFilterExpression = `source.type = "player" and type = "cast" and ability.id IN (${[
  ...remarkableSpellIDs,
].join(", ")})`;

const baseExpressions = [
  deathFilterExpression,
  remarkableSpellFilterExpression,
];

const dungeonExpressionMap: Record<DungeonWithEvents, string[]> = {
  [DungeonIDs.DE_OTHER_SIDE]: dosFilterExpression,
  [DungeonIDs.HALLS_OF_ATONEMENT]: hoaFilterExpression,
  [DungeonIDs.PLAGUEFALL]: pfFilterExpression,
  [DungeonIDs.SANGUINE_DEPTHS]: sdFilterExpression,
  [DungeonIDs.SPIRES_OF_ASCENSION]: soaFilterExpression,
  [DungeonIDs.THEATRE_OF_PAIN]: topFilterExpression,
  [DungeonIDs.THE_NECROTIC_WAKE]: nwFilterExpression,
};

type AffixWithEvents = Exclude<
  Affixes,
  | "Awakened"
  | "Beguiling"
  | "Tyrannical"
  | "Fortified"
  | "Prideful"
  | "Raging"
  | "Infested"
  | "Inspiring"
  | "Reaping"
  | "Skittish"
>;

const affixExpressionMap: Record<AffixWithEvents, string[]> = {
  [Affixes.Sanguine]: sanguineFilterExpression,
  [Affixes.Explosive]: explosiveFilterExpression,
  [Affixes.Grievous]: grievousFilterExpression,
  [Affixes.Necrotic]: necroticFilterExpression,
  [Affixes.Volcanic]: volcanicFilterExpression,
  [Affixes.Bursting]: burstingFilterExpression,
  [Affixes.Spiteful]: spitefulFilterExpression,
  [Affixes.Quaking]: quakingFilterExpression,
  [Affixes.Storming]: stormingFilterExpression,
  [Affixes.Bolstering]: bolsteringFilterExpression,
  [Affixes.Tormented]: tormentedFilterExpression,
};

const generateFilterExpression = ({
  dungeonID,
  affixes,
}: Pick<EventParams, "dungeonID" | "affixes">) => {
  const dungeonExpressions = isDungeonWithEvent(dungeonID)
    ? dungeonExpressionMap[dungeonID]
    : [];

  const affixExpressions = affixes
    .filter((affix): affix is AffixWithEvents => affix in affixExpressionMap)
    .flatMap((affix) => affixExpressionMap[affix]);

  return [...baseExpressions, ...dungeonExpressions, ...affixExpressions]
    .flat()
    .map((part) => `(${part})`)
    .join(" or ");
};

const filterDungeonEvents = (
  allEvents: AnyEvent[],
  dungeonID: DungeonIDs,
  playerMetaInformation: Parameters<typeof getEvents>[1]
) => {
  switch (dungeonID) {
    case DungeonIDs.DE_OTHER_SIDE:
      return allEvents.filter(isDosUrnEvent);
    case DungeonIDs.HALLS_OF_ATONEMENT:
      return allEvents.filter(isHoaGargoyleEvent);
    case DungeonIDs.THE_NECROTIC_WAKE:
      return [
        ...allEvents
          .filter(isNwSpearEvent)
          .reduce<DamageEvent[][]>(nwSpearReducer, [])
          .flatMap((chunk) => reduceEventsByPlayer(chunk, "sourceID")),
        ...allEvents
          .filter(isNwOrbEvent)
          .reduce<DamageEvent[][]>(nwOrbReducer, [])
          .flatMap((chunk) => reduceEventsByPlayer(chunk, "sourceID")),
        ...allEvents.filter(isNwHammerEvent),
        ...allEvents.filter(isNwKyrianOrbDamageEvent),
        ...allEvents.filter(isNwKyrianOrbHealEvent),
      ];
    case DungeonIDs.PLAGUEFALL: {
      const actorIDSet = new Set(
        playerMetaInformation.map((player) => player.actorID)
      );

      const plagueBombDamageEvents = allEvents.filter(
        isPfPlagueBombDamageEvent
      );

      const plagueBombDamageTakenEvents = reduceEventsByPlayer(
        plagueBombDamageEvents.filter((event) =>
          actorIDSet.has(event.targetID)
        ),
        "targetID"
      );
      const plagueBombDamageDoneEvents = plagueBombDamageEvents.filter(
        (event) => !actorIDSet.has(event.targetID)
      );

      const aggregatedPlagueBombDamageEvent =
        plagueBombDamageDoneEvents.length > 0
          ? plagueBombDamageDoneEvents.reduce((acc, event) => {
              if (acc === event) {
                return acc;
              }

              return {
                ...acc,
                amount: acc.amount + event.amount - (event.overkill ?? 0),
              };
            }, plagueBombDamageDoneEvents[0])
          : null;

      return [
        ...allEvents.filter((event) => isPfSlimeDeathEvent(event, actorIDSet)),
        ...allEvents.filter(isPfSlimeBuffEvent),
        ...plagueBombDamageTakenEvents,
        ...(aggregatedPlagueBombDamageEvent
          ? [aggregatedPlagueBombDamageEvent]
          : []),
      ];
    }
    case DungeonIDs.SANGUINE_DEPTHS:
      return [
        ...allEvents.filter(isSdLanternBuffEvent),
        ...allEvents.filter(isSdLanternOpeningEvent),
      ];
    case DungeonIDs.THEATRE_OF_PAIN:
      return allEvents.filter(isTopBannerAuraEvent);
    case DungeonIDs.SPIRES_OF_ASCENSION:
      return (
        allEvents
          .filter(isSoaSpearEvent)
          .reduce<ApplyDebuffEvent[][]>(soaSpearReducer, [])
          // pick only the first event of each chunk,
          // indicating when the spear was used
          .flatMap((chunk) => chunk[0])
      );
    default:
      return [];
  }
};

const filterTormentedEvents = (allEvents: AnyEvent[]) => {
  // powers
  const stygianKingsBarbs = reduceEventsByPlayer(
    allEvents.filter(isStygianKingsBarbsEvent),
    "sourceID"
  );
  const theFifthSkulL = reduceEventsByPlayer(
    allEvents.filter(isTheFifthSkullDamageEvent),
    "sourceID"
  );
  const bottleOfSanguineIchorDamage = reduceEventsByPlayer(
    allEvents.filter(isBottleOfSanguineIchorDamageEvent),
    "sourceID"
  );
  const bottleOfSanguineIchorHeal = reduceEventsByPlayer(
    allEvents.filter(isBottleOfSanguineIchorHealEvent),
    "sourceID"
  );
  const volcanicPlume = reduceEventsByPlayer(
    allEvents.filter(isVolcanicPlumeDamageEvent),
    "sourceID"
  );
  const stoneWard = reduceEventsByPlayer(
    allEvents.filter(isStoneWardEvent),
    "sourceID"
  );

  // lieutenant abilities
  const infernoDamageEvent = reduceEventsByPlayer(
    allEvents.filter(isInfernoDamageEvent),
    "targetID"
  );
  const scorchingBlast = reduceEventsByPlayer(
    allEvents.filter(isScorchingBlastDamageEvent),
    "targetID"
  );
  const soulforgeFlame = reduceEventsByPlayer(
    allEvents.filter(isSoulforgeFlamesDamageEvent),
    "targetID"
  );
  const coldSnap = reduceEventsByPlayer(
    allEvents.filter(isColdSnapDamageEvent),
    "targetID"
  );
  const frostLance = reduceEventsByPlayer(
    allEvents.filter(isFrostLanceDamageEvent),
    "targetID"
  );
  const bitingCold = reduceEventsByPlayer(
    allEvents.filter(isBitingColdDamageEvent),
    "targetID"
  );
  const seismicWave = reduceEventsByPlayer(
    allEvents.filter(isSeismicWaveDamageEvent),
    "targetID"
  );
  const crush = reduceEventsByPlayer(
    allEvents.filter(isCrushDamageEvent),
    "targetID"
  );
  const sever = reduceEventsByPlayer(
    allEvents.filter(isSeverDamageEvent),
    "targetID"
  );
  const raze = reduceEventsByPlayer(
    allEvents.filter(isRazeDamageEvent),
    "targetID"
  );

  return [
    ...stygianKingsBarbs,
    ...bottleOfSanguineIchorDamage,
    ...bottleOfSanguineIchorHeal,
    ...infernoDamageEvent,
    ...volcanicPlume,
    ...scorchingBlast,
    ...soulforgeFlame,
    ...coldSnap,
    ...frostLance,
    ...bitingCold,
    ...crush,
    ...sever,
    ...raze,
    ...theFifthSkulL,
    ...stoneWard,
    ...seismicWave,
  ];
};

const filterAffixEvents = (
  allEvents: AnyEvent[],
  affixes: EventParams["affixes"]
) => {
  const affixSet = new Set(affixes);

  // seasonal
  console.time("filterTormentedEvents");
  const tormentedEvents = affixSet.has(Affixes.Tormented)
    ? filterTormentedEvents(allEvents)
    : [];
  console.timeEnd("filterTormentedEvents");

  // common affixes
  console.time("filter common affixes");
  const hasSanguine = affixSet.has(Affixes.Sanguine);
  const hasQuaking = affixSet.has(Affixes.Quaking);
  const hasNecrotic = affixSet.has(Affixes.Necrotic);
  const hasExplosive = affixSet.has(Affixes.Explosive);

  const explosiveTargetID = hasExplosive
    ? findExplosiveTargetID(allEvents)
    : null;

  console.log({ explosiveTargetID });

  if (hasExplosive && !explosiveTargetID) {
    console.error("could not determine targetID for explosives");
  }

  const storming = affixSet.has(Affixes.Storming)
    ? reduceEventsByPlayer(allEvents.filter(isStormingEvent), "targetID")
    : [];
  const spiteful = affixSet.has(Affixes.Spiteful)
    ? reduceEventsByPlayer(allEvents.filter(isSpitefulDamageEvent), "targetID")
    : [];
  const sanguineDamage = hasSanguine
    ? reduceEventsByPlayer(allEvents.filter(isSanguineDamageEvent), "targetID")
    : [];
  const sanguineHeal = hasSanguine
    ? reduceHealingDoneBySanguine(allEvents.filter(isSanguineHealEvent))
    : [];
  const volcanic = affixSet.has(Affixes.Volcanic)
    ? reduceEventsByPlayer(allEvents.filter(isVolcanicEvent), "targetID")
    : [];
  const quakingDamage = hasQuaking
    ? reduceEventsByPlayer(allEvents.filter(isQuakingDamageEvent), "targetID")
    : [];
  const quakingInterrupts = hasQuaking
    ? allEvents.filter(isQuakingInterruptEvent)
    : [];
  const bolstering = affixSet.has(Affixes.Bolstering)
    ? getHighestBolsteringStack(allEvents.filter(isBolsteringEvent))
    : [];
  const bursting = affixSet.has(Affixes.Bursting)
    ? reduceEventsByPlayer(allEvents.filter(isBurstingEvent), "targetID")
    : [];
  const explosiveDamage = hasExplosive
    ? reduceEventsByPlayer(allEvents.filter(isExplosiveDamageEvent), "targetID")
    : [];
  const explosiveKills =
    hasExplosive && explosiveTargetID
      ? allEvents.filter(createIsExplosiveDeathEvent(explosiveTargetID))
      : [];
  const grievous = affixSet.has(Affixes.Grievous)
    ? reduceEventsByPlayer(allEvents.filter(isGrievousDamageEvent), "targetID")
    : [];
  const necroticStacks = hasNecrotic
    ? getHighestNecroticStack(allEvents.filter(isNecroticStackEvent))
    : [];
  const necroticDamage = hasNecrotic
    ? reduceEventsByPlayer(allEvents.filter(isNecroticDamageEvent), "targetID")
    : [];
  console.timeEnd("filter common affixes");

  return [
    ...tormentedEvents,
    ...storming,
    ...spiteful,
    ...sanguineDamage,
    ...sanguineHeal,
    ...volcanic,
    ...quakingDamage,
    ...quakingInterrupts,
    ...bolstering,
    ...bursting,
    ...explosiveDamage,
    ...explosiveKills,
    ...grievous,
    ...necroticStacks,
    ...necroticDamage,
  ];
};

const filterRemarkableSpellEvents = (allEvents: AnyEvent[]): CastEvent[] => {
  return allEvents.filter(
    (event): event is CastEvent =>
      event.type === "cast" && remarkableSpellIDs.has(event.abilityGameID)
  );
};

const filterPlayerDeathEvents = (
  allEvents: AnyEvent[],
  playerMetaInformation: Parameters<typeof getEvents>[1],
  remarkableSpellEvents: CastEvent[]
): DeathEvent[] => {
  const actorIDSet = new Set(
    playerMetaInformation.map((dataset) => dataset.actorID)
  );

  const hunter = playerMetaInformation.find(
    (player) => player.class === PlayableClass.Hunter
  );

  const deathEvents = allEvents.filter((event): event is DeathEvent => {
    return (
      event.type === "death" &&
      actorIDSet.has(event.targetID) &&
      event.sourceID === -1
    );
  });

  if (!hunter) {
    return deathEvents;
  }

  const hunterDeaths = deathEvents.filter(
    (event) => event.targetID === hunter.actorID
  );

  if (hunterDeaths.length === 0) {
    return deathEvents;
  }

  return deathEvents.filter((event) => {
    const isHunterDeath = hunterDeaths.includes(event);

    if (!isHunterDeath) {
      return true;
    }

    const nextHunterCD = remarkableSpellEvents.find((e) => {
      return e.sourceID === event.targetID && e.timestamp > event.timestamp;
    });

    if (!nextHunterCD) {
      return true;
    }

    // assume a hunter feigned if he used a cd within the next 2 seconds
    // could alternatively be solved by querying
    // hostilityType: Friendlies, dataType: Deaths
    // once separately...
    return nextHunterCD.timestamp - event.timestamp >= 2000;
  });
};

export const getEvents = async (
  params: EventParams,
  playerMetaInformation: { actorID: number; class: PlayableClass }[]
): Promise<{ allEvents: AnyEvent[]; playerDeathEvents: DeathEvent[] }> => {
  const filterExpression = generateFilterExpression({
    dungeonID: params.dungeonID,
    affixes: params.affixes,
  });

  console.log(filterExpression);

  console.time(`getEvents.recursiveGetEvents-${params.reportID}`);
  const allEvents = await recursiveGetEvents({
    startTime: params.startTime,
    endTime: params.endTime,
    reportID: params.reportID,
    filterExpression,
  });
  console.timeEnd(`getEvents.recursiveGetEvents-${params.reportID}`);

  console.time(`filterDungeonEvents-${params.reportID}`);
  const dungeonEvents = filterDungeonEvents(
    allEvents,
    params.dungeonID,
    playerMetaInformation
  );
  console.timeEnd(`filterDungeonEvents-${params.reportID}`);
  console.time(`filterAffixEvents-${params.reportID}`);
  const affixEvents = filterAffixEvents(allEvents, params.affixes);
  console.timeEnd(`filterAffixEvents-${params.reportID}`);
  console.time(`filterProfessionEvents-${params.reportID}`);
  const professionEvents = filterProfessionEvents(allEvents);
  console.timeEnd(`filterProfessionEvents-${params.reportID}`);

  console.time(`remarkableSpellEvents-${params.reportID}`);
  const remarkableSpellEvents = filterRemarkableSpellEvents(allEvents);
  console.timeEnd(`remarkableSpellEvents-${params.reportID}`);

  console.time(`filterPlayerDeathEvents-${params.reportID}`);
  const playerDeathEvents = filterPlayerDeathEvents(
    allEvents,
    playerMetaInformation,
    remarkableSpellEvents
  );
  console.timeEnd(`filterPlayerDeathEvents-${params.reportID}`);

  return {
    allEvents: [
      ...professionEvents,
      ...dungeonEvents,
      ...affixEvents,
      ...remarkableSpellEvents,
      ...playerDeathEvents,
    ].sort((a, b) => a.timestamp - b.timestamp),
    playerDeathEvents,
  };
};
