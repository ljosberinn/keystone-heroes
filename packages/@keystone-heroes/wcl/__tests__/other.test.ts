import type { AnyEvent } from "@keystone-heroes/wcl/queries";

import { filterProfessionEvents } from "../src/queries/events/other";
import dimensionalShifterEvents from "./fixtures/dimensionalShifterEvents.json";
import engineeringSLCastEvents from "./fixtures/engineeringSLCastEvents.json";
import potionOfTheHiddenSpiritEvents from "./fixtures/potionOfTheHiddenSpiritEvents.json";
import spellCastEvents from "./fixtures/spellCastEvents.json";

describe("other", () => {
  test("filterProfessionEvents", () => {
    const events: AnyEvent[] = [
      ...dimensionalShifterEvents,
      ...engineeringSLCastEvents,
      ...potionOfTheHiddenSpiritEvents,
      ...spellCastEvents,
    ].sort((a, b) => a.timestamp - b.timestamp);

    expect(filterProfessionEvents(events)).toMatchSnapshot();
  });

  // events from https://www.warcraftlogs.com/reports/fxq2w3aAW49dHhjb#fight=3

  // events from https://www.warcraftlogs.com/reports/fxq2w3aAW49dHhjb#fight=3
});
