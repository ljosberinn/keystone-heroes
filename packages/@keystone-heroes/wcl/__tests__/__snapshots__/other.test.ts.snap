// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`other filterProfessionEvents 1`] = `
Array [
  Object {
    "abilityGameID": 307195,
    "sourceID": 1859,
    "sourceMarker": 6,
    "targetID": 1859,
    "targetMarker": 6,
    "timestamp": 12294824,
    "type": "applybuff",
  },
  Object {
    "abilityGameID": 321422,
    "sourceID": 4,
    "targetID": 4,
    "timestamp": 12295325,
    "type": "applybuff",
  },
  Object {
    "abilityGameID": 321422,
    "sourceID": 1846,
    "sourceMarker": 2,
    "targetID": 1846,
    "targetMarker": 2,
    "timestamp": 12297167,
    "type": "applybuff",
  },
  Object {
    "abilityGameID": 345130,
    "sourceID": 5,
    "targetID": 103,
    "timestamp": 2826042,
    "type": "cast",
  },
]
`;
